tet.rs/src/tetris/board/my_board.rs

```diff
@@ -0,0 +1,382 @@
+            self.second_bag = self.get_random_bag();
+        }
+
+        if self.queue.is_empty() {
+            for _ in 0..self.second_bag.len() {
+                self.queue.push_back(self.second_bag.pop_front().unwrap());
+            }
+
+            self.second_bag = self.get_random_bag();
+        }
+    }
+
+    pub fn get_queue_at(&self, index: usize) -> PieceType {
+        let size: usize = self.queue.len();
+        if index < size {
+            self.queue[index]
+        } else {
+            self.second_bag[index - size]
+        }
+    }
+
+    fn pop_piece_from_queue(&mut self) -> PieceType {
+        self.ensure_complete_queue(false);
+
+        self.queue.pop_front().unwrap()
+    }
+
+    pub fn spawn_next_piece(&mut self) -> bool {
+        let piece_type: PieceType = self.pop_piece_from_queue();
+
+        self.has_held = false;
+
+        self.spawn_piece(piece_type)
+    }
+
+    fn spawn_piece(&mut self, piece_type: PieceType) -> bool {
+        let spawn_position: PiecePosition = Helper::get_piece_spawn_position(piece_type);
+
+        match self.put_piece_at(piece_type, spawn_position) {
+            true => {
+                self.active_piece_type = piece_type;
+                self.active_piece_position = spawn_position;
+                true
+            }
+            false => false,
+        }
+    }
+
+    fn put_piece_at(&mut self, piece_type: PieceType, piece_position: PiecePosition) -> bool {
+        let px: PositionT = piece_position.get_x();
+        let py: PositionT = piece_position.get_y();
+
+        let piece_matrix: MatrixT =
+            PieceMatrix::get_matrix_for(piece_type, piece_position.get_rotation());
+
+        let piece_size: usize = PieceMatrix::get_piece_size(piece_type);
+
+        let save: BoardT = self.board.clone();
+
+        for y in 0..piece_size {
+            for x in 0..piece_size {
+                if !PieceMatrix::test_bit(piece_matrix, piece_type, x as PositionT, y as PositionT)
+                {
+                    continue;
+                }
+
+                let x: PositionT = x as PositionT;
+                let y: PositionT = y as PositionT;
+
+                if x + px >= BOARD_WIDTH || y + py >= BOARD_HEIGHT * 2 || self.get_cell(x, y) {
+                    self.board = save;
+                    return false;
+                }
+
+                self.set_cell(x + px, y + py, true);
+            }
+        }
+
+        true
+    }
+
+    fn xy_to_index(x: PositionT, y: PositionT) -> usize {
+        // TODO: assert coordinates are valid
+        (y * BOARD_WIDTH + x) as usize
+    }
+
+    pub fn is_valid_move(&self, dx: PositionT, dy: PositionT, dr: PositionT) -> bool {
+        if dx != 0 && dy != 0 || dx != 0 && dr != 0 || dy != 0 && dr != 0 {
+            return false;
+        }
+
+        if dr != 0 {
+            dr.abs() <= 2
+        } else if dx != 0 {
+            dx.abs() == 1
+        } else {
+            dy.abs() == 1
+        }
```

Comment:
```suggestion
        match (dx, dy, dr) {
            (_, _, dr) if dr != 0 => dr.abs() <= 2,
            (dx, _, _) if dx != 0 => dx.abs() == 1,
            (_, dy, _) if dy != 0 => dy.abs() == 1,
            _ => false,
        }

```

Reply:
Replacing `_`s with `0`s to make sure only one value is set.
Also reading through this snippet taugh me you can fit additionnal conditions while matching, thanks !